File: coding-interview-patterns/scala/src/main/scala/patterns/SlidingWindow.scala
package patterns
import scala.collection.mutable
object SlidingWindow {
  def minWindow(s:String, t:String): String = {
    if (s.isEmpty || t.isEmpty) return ""
    val need = t.groupBy(identity).view.mapValues(_.length).toMap
    val have = mutable.Map[Char,Int]().withDefaultValue(0)
    var req=need.size; var formed=0
    var l=0; var best=(Int.MaxValue,0,0)
    for (r <- s.indices){
      val ch=s(r); have(ch)=have(ch)+1
      if (need.contains(ch) && have(ch)==need(ch)) formed+=1
      while (formed==req){
        if (r-l+1 < best._1) best=(r-l+1,l,r)
        val c=s(l); have(c)=have(c)-1
        if (need.contains(c) && have(c)<need(c)) formed-=1
        l+=1
      }
    }
    if (best._1==Int.MaxValue) "" else s.substring(best._2, best._3+1)
  }
  def longestSubstringNoRepeat(s:String): Int = {
    val last = mutable.Map[Char,Int]()
    var l=0; var best=0
    for (r <- s.indices){
      val ch=s(r)
      if (last.contains(ch) && last(ch)>=l) l=last(ch)+1
      last(ch)=r; best=best.max(r-l+1)
    }
    best
  }
  def maxSlidingWindow(nums:Array[Int], k:Int): Array[Int] = {
    val q = scala.collection.mutable.ArrayDeque[Int]()
    val out = scala.collection.mutable.ArrayBuffer[Int]()
    for (i <- nums.indices){
      while (q.nonEmpty && nums(q.last) <= nums(i)) q.removeLast()
      q.append(i)
      if (q.head <= i - k) q.removeHead()
      if (i >= k-1) out.append(nums(q.head))
    }
    out.toArray
  }
}

File: coding-interview-patterns/python/patterns/sliding_window.py
from collections import Counter
def min_window(s: str, t: str) -> str:
    if not s or not t: return ""
    need=Counter(t); have=Counter(); req=len(need); formed=0; l=0; best=(10**9,None,None)
    for r,ch in enumerate(s):
        have[ch]+=1
        if ch in need and have[ch]==need[ch]: formed+=1
        while formed==req:
            if r-l+1<best[0]: best=(r-l+1,l,r)
            c=s[l]; have[c]-=1
            if c in need and have[c]<need[c]: formed-=1
            l+=1
    return "" if best[1] is None else s[best[1]:best[2]+1]

def length_of_longest_substring(s: str) -> int:
    last={}; l=best=0
    for r,ch in enumerate(s):
        if ch in last and last[ch]>=l: l=last[ch]+1
        last[ch]=r; best=max(best,r-l+1)
    return best

from collections import deque
def max_sliding_window(nums, k):
    q=deque(); out=[]
    for i,x in enumerate(nums):
        while q and nums[q[-1]]<=x: q.pop()
        q.append(i)
        if q[0]<=i-k: q.popleft()
        if i>=k-1: out.append(nums[q[0]])
    return out
